post: running graphics apps in Docker (#72)
cloudx-labs/posts: posts/navarroaxel/13.running-graphics-apps-in-docker.md

---
title: 'Running graphic apps in Docker: AWS WorkSpaces'
published: true
description: 'How to run GUI apps using X.Org and Docker in Linux or Windows'
tags: 'tutorial, docker, aws, security'
cover_image: ./assets/gui-apps-docker-cover.jpg
id: 1253471
---

> As a Linux user, I need to connect to an Amazon WorkSpace but the client app can only be run in specific old, but maintained, versions of Ubuntu. I tried with the latest Ubuntu version but it didn't work, so I've created a virtual machine with Ubuntu Focal 20.04 but this setup used 14GB of disk space 😵 only to open a window application. This didn't make any sense. Also this included issues when sharing the clipboard from the host to the VM and then to the virtual desktop hosted in Amazon. 🙄

Reviewer: ```suggestion
As a Linux user, I need to connect to an Amazon WorkSpace but the client app can only be run in specific old, but maintained, versions of Ubuntu. I've tried with the latest Ubuntu version but it haven't worked, so created a virtual machine with Ubuntu Focal 20.04 but this setup used 14GB of disk space 😵 only to open a window application. This didn't make any sense and also included issues when sharing the clipboard from the host to the VM and then to the virtual desktop hosted in Amazon. 🙄
```

> Then, I remembered that we can run graphic applications using Docker. Let's see how to connect a container to the [X Window System](https://en.wikipedia.org/wiki/X_Window_System) of the host.

Reviewer: ```suggestion
Then, I remembered that we can run graphic applications using Docker. Let's see how to connect a container to the host's [X Window System](https://en.wikipedia.org/wiki/X_Window_System).
```

## Creating the Docker image

> The following Dockerfile build the image using Ubuntu Focal 20.04, and install the Amazon WorkSpace client:

Reviewer: ```suggestion
The following Dockerfile builds the image using Ubuntu Focal 20.04, and installs the Amazon WorkSpace client:
```

```Dockerfile
FROM ubuntu:focal

ENV DEBIAN_FRONTEND=noninteractive

RUN apt update && \
  apt install -y libusb-1.0-0 wget xauth && \
  wget -q -O - https://workspaces-client-linux-public-key.s3-us-west-2.amazonaws.com/ADB332E7.asc | apt-key add - && \
  echo "deb [arch=amd64] https://d3nt0h4h6pmmc4.cloudfront.net/ubuntu focal main" > /etc/apt/sources.list.d/amazon-workspaces-clients.list && \
  apt remove -y wget && \
  apt update && \
  apt install -y workspacesclient && \
  touch /root/.Xauthority && \
  rm -rf /var/lib/apt/lists/*
```

> 💡 `libusb-1.0-0` is a dependency of Amazon WorkSpaces.

Reviewer: Why not "💡 `libusb-1.0-0` is an Amazon WorkSpaces' dependency"? 🤔

Author: It installed by default in the Linux Desktop Environment (GNOME, KDE Plasma, etc.). So, the bug is in the WorkSpaces `.deb` file. They should mark this lib as a dependency, but this bug is not present in any graphic Ubuntu installation.

The `DEBIAN_FRONTEND=noninteractive` environment is used to skip the Ubuntu’s prompt asking for our time zone.

Build the image with the following command:

```bash
docker build -t workspace .
```

> The Docker image size is 1.2GB, a reduction of 91% in comparison to the 14GB of the virtual machine.

Reviewer: ```suggestion
The Docker image's size is 1.2GB, a reduction of 91% in comparison to the 14GB of the virtual machine.
```

## Starting the container

Now, we can run the container!

```bash
docker run --name ws --network=host -e DISPLAY -v /tmp/.X11-unix -it workspace
```

> We need to share the `.X11-unix/` directory where the connection [socket](https://en.wikipedia.org/wiki/Unix_domain_socket) is located, and `DISPLAY` is the name of the display of the host.

Reviewer: ```suggestion
We need to share the `.X11-unix/` directory where the connection [socket](https://en.wikipedia.org/wiki/Unix_domain_socket) is located, where `DISPLAY` is the name of the host's display.
```

> The `--network=host` argument makes the container use the same network of the host, and the container does not get its own IP address. This avoids the container being network isolated. 🪛

Reviewer: ```suggestion
The `--network=host` argument makes the container use the same host's network so the container does not get its own IP address. This avoids the container being network isolated. 🪛
```

```bash
xauth add <display_name> <protocol_name> <hex_key>
```

Now you are able to start GUI applications inside the container! 🎉

```bash
/opt/workspacesclient/workspacesclient
```

## Alternative: use VNC instead of X11

We need to make a few changes to run a VNC server in the container:

```Dockerfile
FROM ubuntu:focal

ENV DEBIAN_FRONTEND=noninteractive

RUN apt update && \
  apt install -y libusb-1.0-0 gpg wget x11vnc xvfb && \
  wget -q -O - https://workspaces-client-linux-public-key.s3-us-west-2.amazonaws.com/ADB332E7.asc | apt-key add - && \
  echo "deb [arch=amd64] https://d3nt0h4h6pmmc4.cloudfront.net/ubuntu focal main" > /etc/apt/sources.list.d/amazon-workspaces-clients.list && \
  apt remove -y wget && \
  apt update && \
  apt install -y workspacesclient && \
  rm -rf /var/lib/apt/lists/*

RUN echo "exec /opt/workspacesclient/workspacesclient" > ~/.xinitrc && \
  chmod +x ~/.xinitrc

CMD ["x11vnc", "-create", "-forever"]
```

The `x11vnc` command will start an X11 session in the container launching the application specified in the `~/.xinitrc` file.

```bash
docker build -t workspace .
```

> Then, we can run the container

Reviewer: ```suggestion
Then, we can run the container:
```

> Then, we can run the container

Reviewer: ```suggestion
Then, we can run the container:
```

---
title: 'Running graphic apps in Docker: AWS WorkSpaces'
published: true
description: 'How to run GUI apps using X.Org and Docker in Linux or Windows'
tags: 'tutorial, docker, aws, security'
cover_image: ./assets/gui-apps-docker-cover.jpg
id: 1253471
---

As a Linux user, I need to connect to an Amazon WorkSpace but the client app can only be run in specific old, but maintained, versions of Ubuntu. I tried with the latest Ubuntu version but it didn't work, so I've created a virtual machine with Ubuntu Focal 20.04 but this setup used 14GB of disk space 😵 only to open a window application. This didn't make any sense. Also this included issues when sharing the clipboard from the host to the VM and then to the virtual desktop hosted in Amazon. 🙄
Then, I remembered that we can run graphic applications using Docker. Let's see how to connect a container to the [X Window System](https://en.wikipedia.org/wiki/X_Window_System) of the host.

> > Note: Treating option '--ncp-ciphers' as  '--data-ciphers' (renamed in OpenVPN 2.5).

Reviewer 2: This part seems to be copy & paste from the previous article. Doesn't seem related to this one.

  apt remove -y wget && \
  apt update && \
  apt install -y workspacesclient && \
  rm -rf /var/lib/apt/lists/*

RUN echo "exec /opt/workspacesclient/workspacesclient" > ~/.xinitrc && \
  chmod +x ~/.xinitrc

CMD ["x11vnc", "-create", "-forever"]
```

The `x11vnc` command will start an X11 session in the container launching the application specified in the `~/.xinitrc` file.

```bash
docker build -t workspace .
```

Then, we can run the container

```bash
docker run --name ws -it workspace
```

But how can we know the IP address of the container? Just check it with the following command:

```bash
docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' ws
```

Now you can connect to the VNC server. 🔌

## Conclusion

We can run graphic applications that work in specific Linux distros using Docker 🐋, by connecting directly to our X Window System. This uses less memory and disk space than a VM.

You can get the same in Windows as the host using [Xming](https://sourceforge.net/projects/xming/).

`xauth` could be used in conjunction with SSH to run remote GUI apps 😉 in our local, but that's for another post.

> Alternatively, you can start a VNC server in a container allowing more flexibility either if you use macOS, or if you can't connect using X11 because you're using [Wayland](https://en.wikipedia.org/wiki/Wayland_(display_server_protocol)). But this doesn't file like a native window on your OS.

Reviewer 2: ```suggestion
Alternatively, you can start a VNC server in a container allowing more flexibility either if you use macOS, or if you can't connect using X11 because you're using [Wayland](https://en.wikipedia.org/wiki/Wayland_(display_server_protocol)). But this doesn't feel like a native window on your OS.
```
